File: app/utils/cash_rounding.py
"""Cash payment rounding to a configured currency increment."""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

from app.utils.money import q2


def round_cash_total(
    exact: Decimal,
    increment: Decimal | None,
) -> tuple[Decimal, Decimal]:
    """Return ``(rounded_amount, rounding_difference)`` where difference is rounded - exact."""
    exact_q = q2(exact)
    if increment is None or increment <= Decimal("0"):
        return exact_q, Decimal("0.00")
    inc = increment
    units = (exact_q / inc).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
    rounded = q2(units * inc)
    return rounded, q2(rounded - exact_q)

File: web/src/lib/cashRounding.ts
import Decimal from 'decimal.js';

/** Mirrors `app/utils/cash_rounding.py` (ROUND_HALF_UP to nearest increment). */
export function roundCashTotal(
  exact: Decimal | string | number,
  increment: string | number | null | undefined,
): { rounded: Decimal; roundingDifference: Decimal } {
  const exactDec = new Decimal(exact);
  if (increment == null || increment === '' || new Decimal(increment).lte(0)) {
    const rounded = exactDec.toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
    return { rounded, roundingDifference: new Decimal(0) };
  }
  const inc = new Decimal(increment);
  const units = exactDec.div(inc).toDecimalPlaces(0, Decimal.ROUND_HALF_UP);
  const rounded = units.mul(inc).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
  const roundingDifference = rounded.minus(exactDec.toDecimalPlaces(2, Decimal.ROUND_HALF_UP));
  return { rounded, roundingDifference };
}
